### PR TITLE
Storage adapters (+S3 support)

### PR DESCRIPTION
**Pull request by magnusp**
The intention right now is to add the possibility of using different kinds of backing storage than the current filesystem only approach. RepoService gets extracted to an interface which is bound to the correct implementation based on what the configuration says. The PR will also introduce an S3-backed implementation.

### CODE REVIEW

**Comment by coderpatros** [comment]
Awesome. Sorry it's taking a while to review. I'm on holidays and busy with CycloneDX v1.4 stuff. But this contribution is very much appreciated.

**Comment by magnusp** [comment]
Currently I'm using the Amazon provided S3 client which is, as far as I can see, fully async. While the controller(s) are async, RepoService is only partially async. This makes it somewhat troublesome to just extract an interface and use that for storage implementations. I think it would be beneficial to go full async but I believe that is work that could be done in another PR.

**Comment by coderpatros** [comment]
A full async interface makes sense to me if you do want to include it in this PR.

**Comment by magnusp** [comment]
I've now taken the tests from the filesystem based storage and run them against the new S3RepoService. Also I took the liberty to introduce testcontainers running [minio](https://min.io/) in order to avoid mocks. Also I pulled in [Reactive Extensions](https://github.com/dotnet/reactive) to more easily orchestrate asynchronous execution.

Running in a testcontainer is a great convenience but has an impact on how long tests take to run (an extra second or two right now). I'm unsure if Reactive Extensions should stay as it for the most part only serves as a way to allow the use of `await` in anonymous functions. It does help out a bit when paginated results need to be chunked - see S3RepoService#DeleteAll where the Buffer operator is used. I'm leaning towards letting Reactive Extensions go and instead having async marked methods on the interface instead.

I'll wait for feedback on the work so far before progressing.

docker-compose.yml for setting up a minio instance to test against
```
version: "3.8"

services:
  minio:
    image: "minio/minio"
    ports:
      - '9000:9000'
      - '9001:9001'
    command: server --console-address :9001 /data
```
Minios administrative UI will be hosted on localhost:9001 with the default credentials of minioadmin:minioadmin. Create a bucket called "bomserver" and run the server with the following configuration:
```
{
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Information"
    }
  },
  "AllowedMethods": {
    "Get": true,
    "Post": true,
    "Delete": true
  },
  "Repo": {
    "StorageType": "S3",
    "Options": {
      "Endpoint": "localhost:9000",
      "AccessKey": "minioadmin",
      "SecretKey": "minioadmin",
      "ForcePathStyle": true,
      "UseHttp": true,
      "BucketName": "bomserver"
    }
  }
}
```

File: src/CycloneDX.BomRepoServer/Services/S3RepoService.cs [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+                    await using (var memoryStream = new MemoryStream())
+                    {
+                        Serializer.Serialize(bom, memoryStream);
+                        await _s3Client.PutObjectAsync(new PutObjectRequest
+                        {
+                            BucketName = _bucketName,
+                            Key = fileName,
+                            InputStream = memoryStream,
+                            ContentType = "application/protobuf" // TODO What content type?
+                        });
```
Comment (coderpatros) [inline src/CycloneDX.BomRepoServer/Services/S3RepoService.cs:306]: Apologies, I'm not very familiar with working with the AWS .NET SDK. Does the library set the Content-MD5 header for you? If not do you think it should be set?

Also, the official media types are listed here https://cyclonedx.org/specification/overview/

**Comment by coderpatros** [comment]
It looks pretty good to me. I'd be happy to go full async on the interface. I always planned to and probably should have from the start.

**Review by coderpatros — APPROVED** [review]
My only concern about this is the http default in the example config. So I'm going to change that before releasing this.

**Pull request merged**